File: src/main/java/fr/openent/formulaire/service/FolderService.java
package fr.openent.formulaire.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.entcore.common.user.UserInfos;

public interface FolderService {
    /**
     * List all the folders
     * @param user      user connected
     * @param handler   function handler returning JsonArray data
     */
    void list(UserInfos user, Handler<Either<String, JsonArray>> handler);

    /**
     * Get a folder by id
     * @param folderId  folder identifier
     * @param handler   function handler returning JsonObject data
     */
    void get(String folderId, Handler<Either<String, JsonObject>> handler);

    /**
     * Create a folder
     * @param folder    folder to create
     * @param handler   function handler returning JsonObject data
     */
    void create(JsonObject folder, Handler<Either<String, JsonObject>> handler);

    /**
     * Update a folder
     * @param folderId  folder identifier
     * @param folder    folder to update
     * @param handler   function handler returning JsonObject data
     */
    void update(String folderId, JsonObject folder, Handler<Either<String, JsonObject>> handler);

    /**
     * Delete several folders
     * @param folderIds JsonArray of folder identifier
     * @param handler   function handler returning JsonArray data
     */
    void delete(JsonArray folderIds, Handler<Either<String, JsonArray>> handler);

    /**
     * Move folders
     * @param folderIds JsonArray of folder identifiers to move
     * @param parentId  folder identifier of new parent folder
     * @param handler   function handler returning JsonArray data
     */
    void move(JsonArray folderIds, String parentId, Handler<Either<String, JsonArray>> handler);

    /**
     * Sync number of children for a folder
     * @param user      user connected
     * @param folderId  folder identifier
     * @param handler   function handler returning JsonObject data
     */
    void syncNbChildren(UserInfos user, String folderId, Handler<Either<String, JsonObject>> handler);
}

File: src/main/resources/public/ts/services/FolderService.ts
import {idiom, ng, notify} from 'entcore';
import http from 'axios';
import {Folder} from '../models';
import {DataUtils} from "../utils";

export interface FolderService {
    list() : Promise<any>;
    get(folderId: number) : Promise<any>;
    save(folder: Folder) : Promise<any>;
    create(folder: Folder) : Promise<any>;
    update(folder: Folder) : Promise<any>;
    delete(folders: Folder[]) : Promise<any>;
    move(folders: Folder[], parentId: number) : Promise<any>;
}

export const folderService: FolderService = {

    async list() : Promise<any> {
        try {
            return DataUtils.getData(await http.get('/formulaire/folders'));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.folderService.list'));
            throw err;
        }
    },

    async get(folderId) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/folders/${folderId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.folderService.get'));
            throw err;
        }
    },

    async save(folder) : Promise<any> {
        return folder.id ? await this.update(folder) : await this.create(folder);
    },

    async create(folder) : Promise<any> {
        try {
            return DataUtils.getData(await http.post('/formulaire/folders', folder));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.folderService.create'));
            throw err;
        }
    },

    async update(folder) : Promise<any> {
        try {
            return DataUtils.getData(await http.put(`/formulaire/folders/${folder.id}`, folder));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.folderService.update'));
            throw err;
        }
    },

    async delete(folders) : Promise<any> {
        try {
            return DataUtils.getData(await http.delete('/formulaire/folders', { data: folders } ));
        } catch (e) {
            notify.error(idiom.translate('formulaire.error.folderService.delete'));
            throw e;
        }
    },

    async move(folders, parentId) : Promise<any> {
        try {
            return DataUtils.getData(await http.put(`/formulaire/folders/move/${parentId}`, folders));
        } catch (e) {
            notify.error(idiom.translate('formulaire.error.folderService.move'));
            throw e;
        }
    }
};

export const FolderService = ng.service('FolderService', (): FolderService => folderService);